Issue 730 & 138: implement add all to cart

juzen2003 opened the pull request:
- Initial implementation of add all to the cart 
- Fixes #730 and #138 
- Pending Items:
  - Finalize the UI for the 1st column of table header in browse tab.
    - Current icons are a shopping cart (add all button & tooltips) icon and a toolbox icon (only tooltips). Maybe just one shopping cart icon will look better?
  - Fix the dropdown menu location when hamburger is clicked.
    - Need to fix the calculation of top & left in showMenu in browse.js

juzen2003 committed:
In browse tab, update the hamburger menu to have "Add All Results to Cart"
item, and shrink the padding top & bottom of the menu.
- In base.html (line 273), add a new dropdown item "Add All Results to Cart"
  in the hamburger menu in browse tab.
- In opus.css (line 2337-2342), reduce the vertical spacing of each dropdown
  item in the hamburger menu.

juzen2003 committed:
Add checkbox and tools icon in the first column of table header.
- In browse.js (line 1491-1499), add a checkbox and a toolsIcon in the first
  column of table header in browse tab.
- In opus.css (line 2344-2352), add styling for the first column of table
  header in browse tab.

juzen2003 committed:
Use shopping cart icon as the add all button in the browse table header.
Before performing add all, open a modal to ask for confirmation.
- In base.html (line 367-376), add a confirm modal to ask if user wants
  to add all observations to the cart.
- In opus.css (line 2346-2355), add styling for the first column of table
  header in browse tab.
- In browse.js (line 2248-2294), create addAllToCart to call addall api
  and add all observations to the cart.
- In opus.js (line 812-814), when user click yes in add all confirm modal,
  call addAllToCart to add all observations to the cart.

juzen2003 committed:
Open a modal to ask user's confirmation about adding all observations to
the cart when clicking "Add All Results to Cart" in the hamburger.
- In browse.js (line 443), instead of calling addAllToCart directly, open
  #op-addall-to-cart modal to ask for confirmation when user clicks "Add
  All Results to Cart".

rfrenchseti reviewed (changes requested):
- Add all to cart should only be in the hamburger menu while on the browse tab. It should disappear on the cart tab.
- "Add all results to cart" should be in lower case like the other menu options, not in title case.
- Move "add all" after "add range" in hamburger menu.
- I think we can squeeze more vertical space out of the hamburger menu.
- The dialog asking for confirm of add all should have the number of observations (result count) in it. "Are you sure you want to add all XXX,XXX observations to the cart?" Be sure to include commas.
- We should not allow the user to access the add all option if we know it will fail because there are too many observations. Pass in the settings.MAX_SELECTIONS_ALLOWED parameter as a constant through header.html and then check it in the JS. You can't get it precisely right because you don't know what's already in the cart, but you can at least see if the current result count is > MAX_SELECTIONS_ALLOWED and pop up a dialog that says "There are too many results to add all to the cart. Please reduce the number of results to {MAX_SELECTIONS_ALLOWED} or fewer and try again."
- I think browse.js:addAllToCart should actually be in cart.js. And most of that function duplicates code already in cart.js:toggleInCart. Those functions should somehow be combined or refactored so the code isn't duplicated.

juzen2003 committed:
Make "Add all results to cart" only available in browse tab. Shrink the
vertical spaces of the hamburger menu.
- In base.html (line 274), move "Add all results to cart" below "Start add
  range".
- In opus.css (line 2341-2342), shrink the vertical spaces of the hamburger
  menu.
- In opus.css (line 2350-2353), hide the first column of table header in
  cart tab.
- In browse.js (line 1824-1830), in loadData, only show "Add all results
  to cart" in browse tab.
- In cart.js (line 542-544), make sure "Add all results to cart" is hidden
  in cart tab.

juzen2003 committed:
Add result count in add all modal messages.
- In opus.js (line 793-801), add result count in add all modal messages.

juzen2003 committed:
Add a confirmation modal before add all action. If the result count is
larger than MAX_SELECTIONS_ALLOWED, pop out a warning modal to ask user
to reduce the number of obs. If the result count is less than MAX_SELECTIONS_ALLOWED,
pop out a confirmation modal.
- In base.html (line 377-386), add a warning modal #op-addall-warning-msg.
- In header.html (line 62), pass in MAX_SELECTIONS_ALLOWED to header.html
  which will used in js later.
- In views.py (line 49), pass in settings.MAX_SELECTIONS_ALLOWED to context.
- In opus.css (line 2355), rename op-browse-table-header-addall to
  op-table-header-addall.
- In browse.js (line 529-548), create confirmationBeforeAddAll to pop out
  a warning modal to ask user to reduce the number of obs if the result
  count is larger than MAX_SELECTIONS_ALLOWED, and pop out a confirmation
  modal if the result count is less than MAX_SELECTIONS_ALLOWED.
- In browse.js (line 347-350), add a click event handler to the add all
  icon in browse table view to call confirmationBeforeAddAll.
- In opus.js (line 786-787, 797-799), clean up code style for string
  concatenation with "+".

juzen2003 committed:
Refactor and create editAndHighlightObs to perform add/remove/addrange/
removerange/addall and highlight/de-highlight observations. It will be
used in addAllToCart and toggleInCart.
- In cart.js (line 725), create editAndHighlightObs to handle api call of
  editing and highlighting obs.
- Move addAllToCart from browse.js to cart.js.

rfrenchseti commented on opus/application/apps/ui/templates/ui/base.html:
```diff
@@ -362,6 +363,27 @@ <h3 class='op-header-text'></h3>
+        %}
+        {% include "ui/confirm_modal.html" with
+            id_name="op-addall-warning-msg"
+            confirm_title="Message About Add All Action"
```
Let's change the title to "Failed to Add All Observations"

rfrenchseti commented on opus/application/apps/ui/templates/ui/header.html:
```diff
@@ -59,6 +59,7 @@
         const DEFAULT_SORT_ORDER = '{{ default_sort_order }}';
         const STATIC_URL = '{{ STATIC_URL }}';
         const MAX_SELECTIONS_FOR_DATA_DOWNLOAD = '{{ MAX_SELECTIONS_FOR_DATA_DOWNLOAD }}';
+        const MAX_SELECTIONS_ALLOWED = '{{ max_selections_allowed }}';
```
The line about "MAX_SELECTIONS_FOR_DATA_DOWNLOAD" seems to be old, dead code. The variable isn't even being passed in to the template context. Let's go ahead and remove it.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -523,6 +526,26 @@ var o_browse = {
+        if (o_browse.totalObsCount <= MAX_SELECTIONS_ALLOWED) {
+            $("#op-addall-to-cart").modal("show");
+        } else {
+            let warningMsg = "There are too many results to add all to the cart. " +
```
Change to "There are too many results to add them all to the cart."

rfrenchseti commented on opus/application/static_media/js/cart.js:
```diff
@@ -696,25 +699,52 @@ var o_cart = {
             if (statusData.error) {
                 // if previous error modal is currently open, we store the error message for later displaying
-                if ($("#op-cart-status-error-msg").hasClass("show")) {
+                if ($("#op-cart-status-error-msg").hassClass("show")) {
```
Spelling error hasClass?

rfrenchseti reviewed (changes requested):
- I think the toolbox icon in the table view ends up just being confusing, since you can click on the shopping cart but not on the toolbox. Probably best to just remove it.
- When you're in the middle of adding/removing a range, and do an add all, it leaves you in the middle of doing a range with the green square continuing to pulse. I think it should abort the add range when the add all is confirmed. It looks like you might have explicitly removed this feature around cart.js line 743?

dstopp commented on opus/application/static_media/js/cart.js:
```diff
@@ -539,6 +539,9 @@ var o_cart = {
+        } else {
+            // Make sure "Add all results to cart" is hidden in cart tab.
+            $("#op-obs-menu .dropdown-item[data-action='addall']").addClass("op-hide-element");
         }
```
Curiously, aren't you already doing this in loadData()?

juzen2003 replied:
Yes, but because every time we go to browse tab (call activateBrowseTab, and loadData is called as well in the function), the icon will be shown. And if we switch back to cart tab (from browse tab) again without reloading obs in cart tab, the icon will not be hidden (loadData is not called in activateCartTab). Therefore, we hide the icon here for the case that obs in cart tab is not reloaded.

dstopp reviewed (changes requested):
I did not get a fail message when I started w/9989 items in my cart and tried an "AddAll" w/the search set for 207.  I would expect it to complain that I was exceeding my limit of 10k but i never saw the error.  I know it did error because i see this in my log: 
{"error": "Your request to add all 207 observations to the cart failed. The resulting cart and recycle bin would have more than the maximum (10,000) allowed. None of the observations were added.", "count": 9989, "recycled_count": 0, "reqno"

juzen2003 committed:
Update add all warning modal message, and remove toolbox icon in the first
column of browse table header.
- In base.html (line 379), update add all warning modal message.
- In header.html (line 61), remove unused MAX_SELECTIONS_FOR_DATA_DOWNLOAD.
- In browse.js (line 1517), remove toolbox icon in the first column in the
  browse table header.
- In opus.css (line 2345), remove the styling for the first column in the
  browse table header.
- In cart.js (line 740), fix an error caused by typo.

juzen2003 committed:
Update confirmationBeforeAddAll to pop out an error modal if the obs count
in cart + obs count in browse is larger than MAX_SELECTIONS_ALLOWED. Remove
green border pulse from add range if an add all action happened in the middle
of an add range action.
- In browse.js (line 537-540), update confirmationBeforeAddAll to pop out
  an error modal if the obs count in cart + obs count in browse is larger
  than MAX_SELECTIONS_ALLOWED.
- In cart.js (line 723), call undoRangeSelect in addAllToCart to remove
  green border pulse when add all happened in the middle of add range.

juzen2003 committed:
Update comments in cart.js (line 544).

juzen2003 committed:
Update confirmationBeforeAddAll to pop a warning modal when o_browse.totalObsCount
is larger than MAX_SELECTIONS_ALLOWED. Also align the shopping cart icon
in the first column of browse table header with checkboxes in other rows.
- In opus.css (line 2345-2350), add styling to align shopping cart icon
  with checkboxes in the first column of browse table.
- In browse.js (line 537), modify confirmationBeforeAddAll to display a
  warning modal when o_browse.totalObsCount is larger than MAX_SELECTIONS_ALLOWED.
